File: cpp_concurrent_testing/test/StackNonBlockingTest.cpp
#include "TestConfig.hpp"
#include "RunOperationsStack.hpp"
#include "RunController.hpp"
#include <iostream>
#include <vector>
#include <thread>
#include <memory>
#include <chrono>
#include <atomic>

#include "StructureFactory.hpp"

/**
 * Stack Non-Blocking (Lock-Freedom) Test — mirrors Java StackNonBlockingTest.
 *
 * The ConcurrentDataStructure is pre-instrumented by the Python victim_injector
 * with sleep(5s) after successful CAS in pop().
 * One thread becomes the victim and stalls; the remaining threads must still
 * make progress, proving lock-freedom.
 *
 * Progress is measured by RunController::globalOps. If ops increase
 * while the victim is stalled, lock-freedom is proven.
 *
 * Prints "Sanity Test Passed" or "Sanity Test Failed".
 */
int main(int argc, char** argv) {
    TestConfig config;
    if (!config.parse(argc, argv)) return 1;

    std::cout << "Stack NonBlocking Experiment: Algo:" << config.algo
              << ", Duration:" << config.duration << "s"
              << ", Threads:" << config.numThreads << ", Keys:" << config.keyspaceSize << "\n";

    auto stack = createStack(config.algo);
    if (!stack) {
        std::cerr << "Algorithm " << config.algo << " not implemented yet.\n";
        return 1;
    }

    // Pre-populate so pop() has something to consume
    for (int i = 0; i < config.keyspaceSize / 2; i++) {
        stack->push(i);
    }

    std::vector<int> results(config.numThreads, 0);
    std::vector<std::thread> threads;
    RunController::reset();

    for (int i = 0; i < config.numThreads; i++) {
        threads.emplace_back(RunOperationsStack::benchMarkRun, stack.get(), i,
                             config.keyspaceSize, std::ref(results));
    }

    RunController::startFlag.store(true, std::memory_order_release);

    // Sample globalOps at t=1s and t=duration to confirm progress
    std::this_thread::sleep_for(std::chrono::seconds(1));
    long opsBefore = RunController::globalOps.load(std::memory_order_acquire);
    std::this_thread::sleep_for(std::chrono::seconds(config.duration - 1));
    long opsAfter = RunController::globalOps.load(std::memory_order_acquire);

    RunController::stopFlag.store(true, std::memory_order_release);
    for (auto& t : threads) t.join();

    // Lock-freedom: other threads made progress while victim stalled
    bool passed = (opsAfter - opsBefore) >= (config.numThreads - 1) * 5;

    if (passed) {
        std::cout << "Sanity Test Passed\n";
    } else {
        std::cout << "[StackNonBlockingTest] Progress check failed: opsBefore="
                  << opsBefore << " opsAfter=" << opsAfter << "\n";
        std::cout << "Sanity Test Failed\n";
    }

    return passed ? 0 : 1;
}

File: java_concurrent_testing/src/main/java/com/example/test/StackNonBlockingTest.java
package com.example.test;

import java.util.concurrent.atomic.AtomicInteger;
import com.example.utils.StackADT;

/**
 * Non-Blocking (Lock-Freedom Victim) Test for lock-free Stack implementations.
 *
 * Mechanism:
 * The ConcurrentDataStructure is pre-instrumented with victim-sleep logic
 * anchored on the comment "// Pop victim point" inside pop(). One thread
 * stalls for 10 s inside pop(); the remaining threads must still make
 * measurable progress — proving lock-freedom.
 *
 * Prints "Sanity Test Passed" or "Sanity Test Failed" (parsed by Python
 * runner).
 */
public class StackNonBlockingTest {

    private static int numberOfThreads = 5;
    private static int maxRunningTime = 3;
    private static int keyRange = 100;
    private static StackADT stack;

    public static void main(String[] args) {
        try {
            for (int i = 0; i < args.length - 1; i++) {
                if ("-n".equals(args[i]))
                    numberOfThreads = Integer.parseInt(args[i + 1]);
                if ("-d".equals(args[i]))
                    maxRunningTime = Integer.parseInt(args[i + 1]);
                if ("-k".equals(args[i]))
                    keyRange = Integer.parseInt(args[i + 1]);
            }

            try {
                stack = (StackADT) Class.forName("com.example.Sets.ConcurrentDataStructure").getDeclaredConstructor()
                        .newInstance();
            } catch (Exception e) {
                System.out.println("Sanity Test Failed");
                System.err.println(
                        "[StackNonBlockingTest] Could not instantiate ConcurrentDataStructure: " + e.getMessage());
                return;
            }

            AtomicInteger totalPushed = new AtomicInteger(0);
            AtomicInteger totalPopped = new AtomicInteger(0);
            AtomicInteger invalidPops = new AtomicInteger(0);
            int[] results = new int[numberOfThreads];

            Thread[] threads = new Thread[numberOfThreads];
            for (int i = 0; i < numberOfThreads; i++) {
                threads[i] = new Thread(new RunOperationsStackNonBlocking(
                        stack, i, keyRange, results,
                        totalPushed, totalPopped, invalidPops, true));
            }

            // Pre-populate
            for (int i = 0; i < keyRange / 2; i++) {
                stack.push(i);
                totalPushed.incrementAndGet();
            }

            RunControllerNonBlocking.startFlag = RunControllerNonBlocking.stopFlag = false;
            RunControllerNonBlocking.globalOps.set(0);

            for (Thread t : threads)
                t.start();
            RunControllerNonBlocking.startFlag = true;

            Thread.sleep(1000);
            long opsBefore = RunControllerNonBlocking.globalOps.get();
            Thread.sleep((maxRunningTime - 1) * 1000L);
            long opsAfter = RunControllerNonBlocking.globalOps.get();

            RunControllerNonBlocking.stopFlag = true;
            for (Thread t : threads)
                t.join();

            boolean passed = (opsAfter - opsBefore) >= (numberOfThreads - 1) * 5;

            if (passed) {
                System.out.println("Sanity Test Passed");
            } else {
                System.out.printf("[StackNonBlockingTest] Progress check failed: opsBefore=%d opsAfter=%d%n", opsBefore,
                        opsAfter);
                System.out.println("Sanity Test Failed");
            }
        } catch (Throwable t) {
            System.err.println("[StackNonBlockingTest] FATAL: " + t.getMessage());
            t.printStackTrace(System.err);
            System.out.println("Sanity Test Failed");
        }
    }
}
